<a href="https://colab.research.google.com/github/sam505/Machine_Learning/blob/master/Assignment_9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment (60 pts + 10 pts Extra Credit) ##

**You should update the path below**

import pandas as pd
<br>
pathSuperstore = r'C:\Users\f_sohnoume\OneDrive - UNCG\f_sohnoume (faculty)\Spring 2021\ISM210\Resources\Week 8\Sample - Superstore.xlsx'
<br>
superstoreData = pd.read_excel(pathSuperstore)


* Transform categorical variables into dummies using the function **get_dummies()** (20 pts)
    * Focus on City, State, Region, EmployeeID, Segment, Category, Sub-Category, and Ship Mode.
    * The transformation of each categorical variable is worth *2.5 pts*
* Create a column called "Price" by dividing the column "Sales" with the column "Quantity" (2.5 pts)
    * "Price" represents the average price of the transaction.
* Create a column called "Shipping Time" representing the number of days between the Order Date and the Ship Date (2.5 pts)
    * The difference between two *datetime64* type variables returns a *timedelta* type variable
    * You can use **dt.days** to convert a *timedelta* to the number of days. Also, you can use Google search to see how to convert a *timedelta* to an *int*. Here is an example https://www.kite.com/python/answers/how-to-convert-a-timedelta-column-in-a-pandas-dataframe-to-a-days-column-in-python
* Using **patsy** and **statsmodels**, fit four models that predict *Sales* (25 pts)
    * For each model, choose a different combination of predictors. You can include categorical variables (e.g., State, Region, etc.), and/or numerical variables (e.g., Shipping time, Price, etc.)
    * Each model should have at least 3 predictors
    * Each model is worth *5 pts*
    * Select the best model based on the R-Squared coefficient (5 pts)
        * The best model is the one that has the highest R-Squared coefficient
* Fit a model that predicts *Sales* (5 pts)
    * The model should use standardized variables as predictors. The function that standardizes variables is **standardize()**.
    * The list of predictors is: Discount, Price, and Shipping Time.

**Use the following code to create the training and testing datasets should update the path below**

train = superstoreData[:5000]<br>
test = superstoreData[5000:]


**Use the following code to create the outcome to predict**

train['IsProfitable'] = (train['Profit'] >= 0).astype(int)
test['IsProfitable'] = (test['Profit'] >= 0).astype(int)

* Use the LogisticRegression model of scikit-learn to create a model instance (15 pts)
    * Create a list of predictors. You can include the dummies created earlier in the list of predictors (5 pts)
        * The list of predictors should have at least 5 variables
    * Following the lecture notes, fit the model and calculate the y_predict (5 pts)
    * Following the lecture notes, calculate the y_true and the accuracy percentage (5 pts)


In [ ]:
import pandas as pd
import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf
from sklearn.linear_model import LogisticRegression


pathSuperstore = r'Superstore.xlsx'
superstoreData = pd.read_excel(pathSuperstore)
superstoreData

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


Row ID        Order ID Order Date  ... Discount    Profit EmployeeID
0          1  CA-2017-152156 2017-11-08  ...     0.00   41.9136          5
1          2  CA-2017-152156 2017-11-08  ...     0.00  219.5820          1
2          3  CA-2017-138688 2017-06-12  ...     0.00    6.8714          7
3          4  US-2016-108966 2016-10-11  ...     0.45 -383.0310          9
4          5  US-2016-108966 2016-10-11  ...     0.20    2.5164          6
...      ...             ...        ...  ...      ...       ...        ...
9989    9990  CA-2015-110422 2015-01-21  ...     0.20    4.1028          4
9990    9991  CA-2018-121258 2018-02-26  ...     0.00   15.6332          4
9991    9992  CA-2018-121258 2018-02-26  ...     0.20   19.3932          3
9992    9993  CA-2018-121258 2018-02-26  ...     0.00   13.3200          3
9993    9994  CA-2018-119914 2018-05-04  ...     0.00   72.9480          1

[9994 rows x 22 columns]

# Transform categorical variables into dummies using the function get_dummies() (20 pts)
- ## Focus on City, State, Region, EmployeeID, Segment, Category, Sub-Category, and Ship Mode.
- ## The transformation of each categorical variable is worth 2.5 pts

In [ ]:
dummies = pd.get_dummies(superstoreData.City, prefix='City')
superstoreData_dummies = superstoreData.drop('City', axis=1).join(dummies)
superstoreData_dummies

Row ID        Order ID Order Date  ... City_York City_Yucaipa City_Yuma
0          1  CA-2017-152156 2017-11-08  ...         0            0         0
1          2  CA-2017-152156 2017-11-08  ...         0            0         0
2          3  CA-2017-138688 2017-06-12  ...         0            0         0
3          4  US-2016-108966 2016-10-11  ...         0            0         0
4          5  US-2016-108966 2016-10-11  ...         0            0         0
...      ...             ...        ...  ...       ...          ...       ...
9989    9990  CA-2015-110422 2015-01-21  ...         0            0         0
9990    9991  CA-2018-121258 2018-02-26  ...         0            0         0
9991    9992  CA-2018-121258 2018-02-26  ...         0            0         0
9992    9993  CA-2018-121258 2018-02-26  ...         0            0         0
9993    9994  CA-2018-119914 2018-05-04  ...         0            0         0

[9994 rows x 552 columns]

In [ ]:
dummies = pd.get_dummies(superstoreData_dummies.State, prefix='State')
superstoreData_dummies = superstoreData_dummies.drop('State', axis=1).join(dummies)
superstoreData_dummies

Row ID        Order ID  ... State_Wisconsin State_Wyoming
0          1  CA-2017-152156  ...               0             0
1          2  CA-2017-152156  ...               0             0
2          3  CA-2017-138688  ...               0             0
3          4  US-2016-108966  ...               0             0
4          5  US-2016-108966  ...               0             0
...      ...             ...  ...             ...           ...
9989    9990  CA-2015-110422  ...               0             0
9990    9991  CA-2018-121258  ...               0             0
9991    9992  CA-2018-121258  ...               0             0
9992    9993  CA-2018-121258  ...               0             0
9993    9994  CA-2018-119914  ...               0             0

[9994 rows x 600 columns]

In [ ]:
dummies = pd.get_dummies(superstoreData_dummies.Region, prefix='Region')
superstoreData_dummies = superstoreData_dummies.drop('Region', axis=1).join(dummies)
superstoreData_dummies

Row ID        Order ID Order Date  ... Region_East Region_South Region_West
0          1  CA-2017-152156 2017-11-08  ...           0            1           0
1          2  CA-2017-152156 2017-11-08  ...           0            1           0
2          3  CA-2017-138688 2017-06-12  ...           0            0           1
3          4  US-2016-108966 2016-10-11  ...           0            1           0
4          5  US-2016-108966 2016-10-11  ...           0            1           0
...      ...             ...        ...  ...         ...          ...         ...
9989    9990  CA-2015-110422 2015-01-21  ...           0            1           0
9990    9991  CA-2018-121258 2018-02-26  ...           0            0           1
9991    9992  CA-2018-121258 2018-02-26  ...           0            0           1
9992    9993  CA-2018-121258 2018-02-26  ...           0            0           1
9993    9994  CA-2018-119914 2018-05-04  ...           0            0           1

[9994 rows x 603 columns]

In [ ]:
dummies = pd.get_dummies(superstoreData_dummies.EmployeeID, prefix='EmployeeID')
superstoreData_dummies = superstoreData_dummies.drop('EmployeeID', axis=1).join(dummies)
superstoreData_dummies

Row ID        Order ID  ... EmployeeID_8 EmployeeID_9
0          1  CA-2017-152156  ...            0            0
1          2  CA-2017-152156  ...            0            0
2          3  CA-2017-138688  ...            0            0
3          4  US-2016-108966  ...            0            1
4          5  US-2016-108966  ...            0            0
...      ...             ...  ...          ...          ...
9989    9990  CA-2015-110422  ...            0            0
9990    9991  CA-2018-121258  ...            0            0
9991    9992  CA-2018-121258  ...            0            0
9992    9993  CA-2018-121258  ...            0            0
9993    9994  CA-2018-119914  ...            0            0

[9994 rows x 611 columns]

In [ ]:
dummies = pd.get_dummies(superstoreData_dummies.Segment, prefix='Segment')
superstoreData_dummies = superstoreData_dummies.drop('Segment', axis=1).join(dummies)
superstoreData_dummies

Row ID        Order ID  ... Segment_Corporate Segment_Home Office
0          1  CA-2017-152156  ...                 0                   0
1          2  CA-2017-152156  ...                 0                   0
2          3  CA-2017-138688  ...                 1                   0
3          4  US-2016-108966  ...                 0                   0
4          5  US-2016-108966  ...                 0                   0
...      ...             ...  ...               ...                 ...
9989    9990  CA-2015-110422  ...                 0                   0
9990    9991  CA-2018-121258  ...                 0                   0
9991    9992  CA-2018-121258  ...                 0                   0
9992    9993  CA-2018-121258  ...                 0                   0
9993    9994  CA-2018-119914  ...                 0                   0

[9994 rows x 613 columns]

In [ ]:
dummies = pd.get_dummies(superstoreData_dummies["Sub-Category"], prefix='Sub_Category')
superstoreData_dummies = superstoreData_dummies.drop('Sub-Category', axis=1).join(dummies)
superstoreData_dummies

Row ID        Order ID  ... Sub_Category_Supplies Sub_Category_Tables
0          1  CA-2017-152156  ...                     0                   0
1          2  CA-2017-152156  ...                     0                   0
2          3  CA-2017-138688  ...                     0                   0
3          4  US-2016-108966  ...                     0                   1
4          5  US-2016-108966  ...                     0                   0
...      ...             ...  ...                   ...                 ...
9989    9990  CA-2015-110422  ...                     0                   0
9990    9991  CA-2018-121258  ...                     0                   0
9991    9992  CA-2018-121258  ...                     0                   0
9992    9993  CA-2018-121258  ...                     0                   0
9993    9994  CA-2018-119914  ...                     0                   0

[9994 rows x 629 columns]

In [ ]:
dummies = pd.get_dummies(superstoreData_dummies["Ship Mode"], prefix='Ship_Mode')
superstoreData_dummies = superstoreData_dummies.drop('Ship Mode', axis=1).join(dummies)
superstoreData_dummies

Row ID        Order ID  ... Ship_Mode_Second Class Ship_Mode_Standard Class
0          1  CA-2017-152156  ...                      1                        0
1          2  CA-2017-152156  ...                      1                        0
2          3  CA-2017-138688  ...                      1                        0
3          4  US-2016-108966  ...                      0                        1
4          5  US-2016-108966  ...                      0                        1
...      ...             ...  ...                    ...                      ...
9989    9990  CA-2015-110422  ...                      1                        0
9990    9991  CA-2018-121258  ...                      0                        1
9991    9992  CA-2018-121258  ...                      0                        1
9992    9993  CA-2018-121258  ...                      0                        1
9993    9994  CA-2018-119914  ...                      1                        0

[9994 rows x 632 columns]

# Create a column called "Price" by dividing the column "Sales" with the column "Quantity" (2.5 pts)
- ## "Price" represents the average price of the transaction.

In [ ]:
superstoreData_dummies["Price"] = superstoreData_dummies["Sales"] / superstoreData_dummies["Quantity"]
superstoreData_dummies

Row ID        Order ID  ... Ship_Mode_Standard Class     Price
0          1  CA-2017-152156  ...                        0  130.9800
1          2  CA-2017-152156  ...                        0  243.9800
2          3  CA-2017-138688  ...                        0    7.3100
3          4  US-2016-108966  ...                        1  191.5155
4          5  US-2016-108966  ...                        1   11.1840
...      ...             ...  ...                      ...       ...
9989    9990  CA-2015-110422  ...                        0    8.4160
9990    9991  CA-2018-121258  ...                        1   45.9800
9991    9992  CA-2018-121258  ...                        1  129.2880
9992    9993  CA-2018-121258  ...                        1    7.4000
9993    9994  CA-2018-119914  ...                        0  121.5800

[9994 rows x 633 columns]

# Create a column called "Shipping Time" representing the number of days between the Order Date and the Ship Date (2.5 pts)
- ## The difference between two datetime64 type variables returns a timedelta type variable
- ## You can use dt.days to convert a timedelta to the number of days. Also, you can use Google search to see how to convert a timedelta to an int. Here is an example https://www.kite.com/python/answers/how-to-convert-a-timedelta-column-in-a-pandas-dataframe-to-a-days-column-in-python

In [ ]:
superstoreData_dummies['Shipping_Time'] = (superstoreData_dummies["Ship Date"]-superstoreData_dummies["Order Date"]).dt.days
superstoreData_dummies

Row ID        Order ID  ...     Price Shipping_Time
0          1  CA-2017-152156  ...  130.9800             3
1          2  CA-2017-152156  ...  243.9800             3
2          3  CA-2017-138688  ...    7.3100             4
3          4  US-2016-108966  ...  191.5155             7
4          5  US-2016-108966  ...   11.1840             7
...      ...             ...  ...       ...           ...
9989    9990  CA-2015-110422  ...    8.4160             2
9990    9991  CA-2018-121258  ...   45.9800             5
9991    9992  CA-2018-121258  ...  129.2880             5
9992    9993  CA-2018-121258  ...    7.4000             5
9993    9994  CA-2018-119914  ...  121.5800             5

[9994 rows x 634 columns]

In [ ]:
superstoreData['Shipping_Time'] = (superstoreData["Ship Date"]-superstoreData["Order Date"]).dt.days
superstoreData

Row ID        Order ID Order Date  ...    Profit EmployeeID Shipping_Time
0          1  CA-2017-152156 2017-11-08  ...   41.9136          5             3
1          2  CA-2017-152156 2017-11-08  ...  219.5820          1             3
2          3  CA-2017-138688 2017-06-12  ...    6.8714          7             4
3          4  US-2016-108966 2016-10-11  ... -383.0310          9             7
4          5  US-2016-108966 2016-10-11  ...    2.5164          6             7
...      ...             ...        ...  ...       ...        ...           ...
9989    9990  CA-2015-110422 2015-01-21  ...    4.1028          4             2
9990    9991  CA-2018-121258 2018-02-26  ...   15.6332          4             5
9991    9992  CA-2018-121258 2018-02-26  ...   19.3932          3             5
9992    9993  CA-2018-121258 2018-02-26  ...   13.3200          3             5
9993    9994  CA-2018-119914 2018-05-04  ...   72.9480          1             5

[9994 rows x 23 columns]

# Using patsy and statsmodels, fit four models that predict Sales (25 pts)
- ## For each model, choose a different combination of predictors. You can include categorical variables (e.g., State, Region, etc.), and/or numerical variables (e.g., Shipping time, Price, etc.)
- ## Each model should have at least 3 predictors
- ## Each model is worth 5 pts
- ## Select the best model based on the R-Squared coefficient (5 pts)
  - ### The best model is the one that has the highest R-Squared coefficient

In [ ]:
y, X = patsy.dmatrices('Sales ~ Discount + Quantity + State', superstoreData)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.049
Model:                            OLS   Adj. R-squared:                  0.045
Method:                 Least Squares   F-statistic:                     10.35
Date:                Wed, 03 Nov 2021   Prob (F-statistic):           1.32e-76
Time:                        22:48:48   Log-Likelihood:                -78238.
No. Observations:                9994   AIC:                         1.566e+05
Df Residuals:                    9943   BIC:                         1.569e+05
Df Model:                          50                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
Intercept     

In [ ]:
y, X = patsy.dmatrices('Sales ~ Discount + Quantity + Region + Shipping_Time', superstoreData)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.042
Model:                            OLS   Adj. R-squared:                  0.041
Method:                 Least Squares   F-statistic:                     72.16
Date:                Wed, 03 Nov 2021   Prob (F-statistic):           2.00e-88
Time:                        22:48:48   Log-Likelihood:                -78279.
No. Observations:                9994   AIC:                         1.566e+05
Df Residuals:                    9987   BIC:                         1.566e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          39.8865     22.742     

In [ ]:
y, X = patsy.dmatrices('Sales ~ State + Category + EmployeeID + Country', superstoreData)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.060
Model:                            OLS   Adj. R-squared:                  0.055
Method:                 Least Squares   F-statistic:                     12.40
Date:                Wed, 03 Nov 2021   Prob (F-statistic):           1.12e-97
Time:                        22:48:49   Log-Likelihood:                -78183.
No. Observations:                9994   AIC:                         1.565e+05
Df Residuals:                    9942   BIC:                         1.568e+05
Df Model:                          51                                         
Covariance Type:            nonrobust                                         
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
Intercept     

In [ ]:
y, X = patsy.dmatrices('Sales ~  Quantity + Discount + Profit + Shipping_Time', superstoreData)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.264
Model:                            OLS   Adj. R-squared:                  0.264
Method:                 Least Squares   F-statistic:                     895.2
Date:                Wed, 03 Nov 2021   Prob (F-statistic):               0.00
Time:                        22:48:49   Log-Likelihood:                -76960.
No. Observations:                9994   AIC:                         1.539e+05
Df Residuals:                    9989   BIC:                         1.540e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept       -10.2835     16.471     -0.624

**The best model uses numerical variabes only and achieved an R-squared coefficient of 0.264**

# Fit a model that predicts Sales (5 pts)
- ## The model should use standardized variables as predictors. The function that standardizes variables is standardize().
- ## The list of predictors is: Discount, Price, and Shipping Time.

In [ ]:
y, X = patsy.dmatrices('Sales ~  standardize(Discount) + standardize(Price) + standardize(Shipping_Time)', superstoreData_dummies)
model = sm.OLS(y, X)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.791
Model:                            OLS   Adj. R-squared:                  0.791
Method:                 Least Squares   F-statistic:                 1.260e+04
Date:                Wed, 03 Nov 2021   Prob (F-statistic):               0.00
Time:                        22:48:49   Log-Likelihood:                -70668.
No. Observations:                9994   AIC:                         1.413e+05
Df Residuals:                    9990   BIC:                         1.414e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
Intercept           

# Use the following code to create the training and testing datasets should update the path below

train = superstoreData[:5000]
test = superstoreData[5000:]

Use the following code to create the outcome to predict

train['IsProfitable'] = (train['Profit'] >= 0).astype(int) test['IsProfitable'] = (test['Profit'] >= 0).astype(int)

In [ ]:
train = superstoreData_dummies[:5000]
test = superstoreData_dummies[5000:]

In [ ]:
train['IsProfitable'] = (train['Profit'] >= 0).astype(int) 
test['IsProfitable'] = (test['Profit'] >= 0).astype(int)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


# Use the LogisticRegression model of scikit-learn to create a model instance (15 pts)
- ## Create a list of predictors. You can include the dummies created earlier in the list of predictors (5 pts)
  - ## The list of predictors should have at least 5 variables
- ## Following the lecture notes, fit the model and calculate the y_predict (5 pts)
- ## Following the lecture notes, calculate the y_true and the accuracy percentage (5 pts)

In [ ]:
train.isnull().sum()

Row ID                      0
Order ID                    0
Order Date                  0
Ship Date                   0
Customer ID                 0
                           ..
Ship_Mode_Second Class      0
Ship_Mode_Standard Class    0
Price                       0
Shipping_Time               0
IsProfitable                0
Length: 635, dtype: int64

In [ ]:
test.isnull().sum()

Row ID                      0
Order ID                    0
Order Date                  0
Ship Date                   0
Customer ID                 0
                           ..
Ship_Mode_Second Class      0
Ship_Mode_Standard Class    0
Price                       0
Shipping_Time               0
IsProfitable                0
Length: 635, dtype: int64

In [ ]:
impute_value = train['Postal Code'].median()
train['Postal Code'] = train['Postal Code'].fillna(impute_value)
test['Postal Code'] = test['Postal Code'].fillna(impute_value)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
predictors = ['Price', 'Discount', 'Shipping_Time', 'Quantity', 'Region_Central', 'Region_East', 'Region_South', 'Region_West']
X_train = train[predictors].values
X_test = test[predictors].values
y_train = train['IsProfitable'].values

In [ ]:
X_train[:5]

array([[1.309800e+02, 0.000000e+00, 3.000000e+00, 2.000000e+00,
        0.000000e+00, 0.000000e+00, 1.000000e+00, 0.000000e+00],
       [2.439800e+02, 0.000000e+00, 3.000000e+00, 3.000000e+00,
        0.000000e+00, 0.000000e+00, 1.000000e+00, 0.000000e+00],
       [7.310000e+00, 0.000000e+00, 4.000000e+00, 2.000000e+00,
        0.000000e+00, 0.000000e+00, 0.000000e+00, 1.000000e+00],
       [1.915155e+02, 4.500000e-01, 7.000000e+00, 5.000000e+00,
        0.000000e+00, 0.000000e+00, 1.000000e+00, 0.000000e+00],
       [1.118400e+01, 2.000000e-01, 7.000000e+00, 2.000000e+00,
        0.000000e+00, 0.000000e+00, 1.000000e+00, 0.000000e+00]])

In [ ]:
model = LogisticRegression()
model.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
y_predict = model.predict(X_test)
y_true = test['IsProfitable'].values
(y_true == y_predict).mean() * 100

93.03163796555867